### DISCUSSION TITLE
Installing BTCPAY server via subfolder/subdirectory

### DISCUSSION THREAD

**Discussion by mar101xy** [btcpayserver/btcpayserver#6450] [Ideas / Feature Requests]
I’ve noticed that almost all users install their BTCPay Server on a domain or subdomain, such as:

```
    mydomain.com
    btcpay.mydomain.com
```

I haven’t seen anyone install it in a subfolder. From what I’ve discovered, this might be because it’s not possible. I tried via the server settings "domain name"

![btcpay1](https://github.com/user-attachments/assets/e6b69ede-f36e-4cf5-9608-4f9084d0816d)

I would install it directly on the domain, but I already have a website there. As a result, I’d like to install it in a subfolder, like:

`    mydomain.com/btcpay`

The reason for this preference is that when I create a Lightning Address using the nostr plugin, it includes the subdomain, for example:

`    username@btcpay.mydomain.com`

I’d like to keep it simpler, ideally as:

`    username@mydomain.com`

![btcpay2](https://github.com/user-attachments/assets/d317b79c-fcc6-42e7-9fec-53413b5622c1)

I think other users would also agree and would love a simpler lightning address then with the subdomain on the lightning address

**Comment by dennisreimann** [comment]
There is [the `BTCPAY_ROOTPATH` option](https://github.com/btcpayserver/btcpayserver-docker/blob/master/Production/README.md), have you tried this?

**Comment by mar101xy** [comment]
> There is [the `BTCPAY_ROOTPATH` option](https://github.com/btcpayserver/btcpayserver-docker/blob/master/Production/README.md), have you tried this?

I ended up trying out your suggestion and it worked. I used these variables during installation.


export BTCPAY_PROTOCOL=https
export BTCPAY_HOST=mydomain.com
export BTCPAY_ROOTPATH=/btcpay
export NBITCOIN_NETWORK=mainnet
export BTCPAYGEN_CRYPTO1=btc
export REVERSEPROXY_HTTP_PORT=8080
export REVERSEPROXY_HTTPS_PORT=8443
export BTCPAYGEN_ADDITIONAL_FRAGMENTS="opt-save-storage-xs"
export BTCPAYGEN_LIGHTNING=lnd
export ACME_CA_URI=production
. ./btcpay-setup.sh -i



Now i access my btcpay at:
`mydomain.com/btcpay`


I seem to have found an issue with RTL.
When i access RTL via the admin panel

example:
`https://mydomain.com/rtl/api/authenticate/cookie?access-key=d5b2009001564a7a83261780e31fcabad68cjs72jda8b6575534697e7eb54caaf1886cabbc722cda347317c1bd3725069693923b222523js7c7sh347dhf7shed7dhe`

i get the message:
404 Not Found
nginx version



here is the start of my nginx file
I already have nginx installed on my server because i have a different website on it. 
i am not using the nginx container that comes with btcpay in docker because it has ssl issues so i bypassed the nginx docker container and i point the nginx file to the generated_btcpayserver_1 container

start of my nginx file
```
# Global context
error_log /var/log/nginx/btcpay_debug.log debug;

server {
    listen 80;
    server_name mydomain.com www.mydomain.com;
    return 301 https://$host$request_uri;
}

server {
    listen 443 ssl;
    server_name mydomain.com www.mydomain.com;

    # SSL certificates
    ssl_certificate /etc/letsencrypt/live/mydomain.com/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/mydomain.com/privkey.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;

    # Security headers
    add_header X-Content-Type-Options nosniff;
    add_header Content-Security-Policy "frame-ancestors 'self' https://mydomain.com";
    add_header Strict-Transport-Security "max-age=31536000; includeSubDomains; preload" always;

    # Main site (root)
    root /var/www/mydomain;
    index index.php index.html;

    # Proxy BTCPay traffic
    location /btcpay/ {
        proxy_pass http://172.18.0.10:49392;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-Ssl on;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";  # WebSocket handling
        proxy_buffering off;
        proxy_redirect off;

    # Handle static files correctly (CSS, JS, images)
    location ~* ^/btcpay/(.*\.(css|js|ico|png|jpg|jpeg|svg|woff|woff2|ttf|otf|eot))$ {
        proxy_pass http://172.18.0.10:49392;  # BTCPay container IP:Port for static files
        expires max;
        log_not_found off;
    }
}

    # Proxy RTL traffic
    location /btcpay/rtl/ {
        proxy_pass http://172.18.0.9:3000;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-Ssl on;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";  # WebSocket handling
        proxy_buffering off;
        proxy_redirect off;
    }

    # gRPC Configuration for LND
    location ~* ^/btcpay/(lnrpc|routerrpc|verrpc|walletrpc)\. {
        grpc_read_timeout 6000s;
        grpc_send_timeout 6000s;
        grpc_pass grpcs://172.18.0.7:10009;
    }

    # LND REST API
    location /btcpay/lnd-rest/btc/ {
        rewrite ^/btcpay/lnd-rest/btc/(.*) /$1 break;
        proxy_pass http://172.18.0.7:8080/;
    }

    # Handle 404 for missing static assets
#    error_page 404 /btcpay/404.html;

    # Static file fallback (ensure static files like images, CSS, and JS are handled)
     location ~* ^/btcpay/(.*\.(css|js|ico|png|jpg|jpeg|svg|woff|woff2|ttf|otf|eot))$ {
    proxy_pass http://172.18.0.10:49392;  # BTCPay container IP:Port for static content
    expires max;
    log_not_found off;
}
```

```
sudo docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' generated_btcpayserver_1
sudo docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' generated_lnd_bitcoin_rtl_1
sudo docker inspect -f '{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' btcpayserver_lnd_bitcoin
172.18.0.10
172.18.0.9
172.18.0.7

```


![issue](https://github.com/user-attachments/assets/e47b7dee-a364-4e08-a992-766b969840ea)

**Reply by dennisreimann** [reply]
Would this URL work (with freshly generated credentials) if you insert the /btcpay like `https://mydomain.com/btcpay/rtl/api/authenticate/cookie?access-key=NEW_KEY`?

**Reply by mar101xy** [reply]
It does not work. I get a white page.

![whitepage](https://github.com/user-attachments/assets/b30990f1-8fcb-47f8-9d0e-2a4d35891092)

I inspect element and i saw this:

```
Cookie warnings:
1. Cookie “_csrf” does not have a proper “SameSite” attribute value. This may cause the cookie to be treated as “Lax” in the future. Consider adding the “SameSite=None; Secure” attribute.
2. Cookie “XSRF-TOKEN” does not have a proper “SameSite” attribute value. This may cause the cookie to be treated as “Lax” in the future. Consider adding the “SameSite=None; Secure” attribute.
3. Cookie “connect.sid” does not have a proper “SameSite” attribute value. This may cause the cookie to be treated as “Lax” in the future. Consider adding the “SameSite=None; Secure” attribute.

Resource loading issues:
1. Loading module blocked due to a disallowed MIME type: “runtime.js”.
2. CSS resource blocked due to MIME type mismatch: “styles.css”.
3. Module blocked due to a disallowed MIME type: “polyfills.js” and “main.js”.

Font loading issues:
1. Downloadable font failed: “Roboto-Regular.woff”.

```


i asked chatgpt and it told me to put more things in the nginx file. We tried adding many different things inside the nginx file but none work.


I reverted back to using the root site to see if RTL would work and yes RTL works. I dont even need all the stuff i have now inside the nginx file with i use the root for my btcpay.


export BTCPAY_PROTOCOL=https
export BTCPAY_HOST=mydomain.com
export NBITCOIN_NETWORK=mainnet
export BTCPAYGEN_CRYPTO1=btc
export REVERSEPROXY_HTTP_PORT=8080
export REVERSEPROXY_HTTPS_PORT=8443
export BTCPAYGEN_ADDITIONAL_FRAGMENTS="opt-save-storage-xs"
export BTCPAYGEN_LIGHTNING=lnd
export ACME_CA_URI=production
. ./btcpay-setup.sh -i


then i reverted back to 
` export BTCPAY_ROOTPATH=/btcpay`

and RTL won't work.

I also tried a subdomain and RTL works.

**Reply by dennisreimann** [reply]
Where is the nginx config with the `location /btcpay/rtl/` directive coming from? Is this something you've created yourself? I've checked [our nginx template](https://github.com/btcpayserver/btcpayserver-docker/blob/master/Production/nginx.tmpl#L49) and the `BTCPAY_ROOTPATH` doesn't get incorporated for the external services. Can you try rewriting the RTL location to  `location /rtl/` and try to stay as close as possible to our template?

**Reply by mar101xy** [reply]
This is currently my setup. Can you tell me if it is correct? I am confused on how the server should be setup since btcpay is not the only domain on the server. 

![mysetup-](https://github.com/user-attachments/assets/202f0737-83b0-402f-8871-aedcee1b9fdf)


The server(it has nginx installation on the server)
The server has 3 domains on it already. 
btcpay will be added to the 3rd domain in /btcpay sub directory

the nginx on the server has the conf files in this location:

```
nano /etc/nginx/sites-available/domain1
nano /etc/nginx/sites-available/domain2
nano /etc/nginx/sites-available/domain3
```

those files point to the domains

the last file
`nano /etc/nginx/sites-available/domain3`

inside the nginx file, is the file that i sent you already. It is pointing to the IP's of these 3 containers:

`generated_btcpayserver_1 `
`generated_lnd_bitcoin_rtl_1`
`btcpayserver_lnd_bitcoin`



when i do that, i can access the btcpay domain at:
`mydomain.com/btcpay`



is this the correct way to setup btcpay? 
Is pointing the server nginx file to the 3 containers the correct way to do it? 

or, am i supposed to point the server nginx file to one of the nginx docker container?

`letsencrypt-nginx-proxy-companion`
`nginx-gen`
`nginx`  



i tried pointing the nginx server file to:
`nginx`

container, but it gives me ssl not secure when i visit the domain. 

I thought if i pointed the nginx server file to the nginx container, it would save me from having all the stuff inside the nginx server file.


There is already 2 other domains on the server so i cannot setup btcpay alone. 

i tried subdomain & root domain and everything works. 

With the current 
`mydomain.com/btcpay `

btcpay work but the RTL doesnt work. 
what is even more weird is. if i access my btcpay site this way, btcpay & RTL work

```
https://mydomain.com:8443/btcpay
https://mydomain.com:8443/rtl/lnd/home
```



I used chatgpt to help me with the nginx server file because its too hard and i only understand 40% of what im doing. It wrote it for me. 

![notworks](https://github.com/user-attachments/assets/fac6aa80-a710-462a-8389-70bbec003841)

![works-8443](https://github.com/user-attachments/assets/b02329dd-87f9-472f-b77c-9aa771510bad)

**Reply by dennisreimann** [reply]
Well, this is really involved and the closest we have as detailed description for cases like yours is this doc:
https://docs.btcpayserver.org/FAQ/Deployment/#can-i-use-an-existing-nginx-server-as-a-reverse-proxy-with-ssl-termination

**Comment by mar101xy** [comment]
I seem to have got it working. Petzsch sent this link on telegram:
https://gist.github.com/petzsch/32304778af54fda3a3788fafd84c75ce

he is using apache, and i am using nginx. I tried apache in the past and couldn't do it but that was before the link.
I think i could get it working. 

The issue is during installation of btcpay you need to pick the correct variables. I was confused on which variables to use. The variables below seem to work when using /btcpay sub dir and using nginx behind btcpay.

I didn't even need to add all the nginx configurations in the nginx file like i was doing before. All you need is 10 lines in the nginx configurations. RTL is working as well. 

The installation is a bit confusing because you need to know which variables to use. I will setup a lightning channel and see if my lightning address works as my original goal is to use a `username@mydomain.com` lightning address

don't ask me to explain the stuff below. im still trying to understand it.

hope this info helps others

```
export BTCPAY_HOST="mydomain.com"
export BTCPAY_ROOTPATH=/btcpay
export NBITCOIN_NETWORK="mainnet"
export BTCPAYGEN_CRYPTO1="btc"
export BTCPAYGEN_ADDITIONAL_FRAGMENTS="opt-save-storage-xs"
export BTCPAYGEN_LIGHTNING="lnd"
export BTCPAY_ENABLE_SSH="true"
export REVERSEPROXY_HTTP_PORT="127.0.0.1:905"
export NOREVERSEPROXY_HTTP_PORT="9191"
export BTCPAYGEN_EXCLUDE_FRAGMENTS="$BTCPAYGEN_EXCLUDE_FRAGMENTS;nginx-https"
export BTCPAYGEN_REVERSEPROXY="none"
. ./btcpay-setup.sh -i
```


nginx configuration

```
    # Proxy to BTC container
    location /btcpay/ {
        proxy_pass http://127.0.0.1:9191;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

}

    # Proxy to RTL container
        location /rtl/ {
                proxy_pass http://192.168.0.2:3000/rtl/;

}
```

**Reply by remcoros** [reply]
lightning addresses don't support subdirectories, the different lnurl/nostr specifications all rely on a "<domain.tld>/.well-known/..." path.

**Comment by mar101xy** [comment]
btcpay admin panel gives me a warning in the lightning settings. In the setting it say it won't work but then it says i can re-direct it. 

I setup a re-direct in nginx then did a cmd test and the re-direct worked but i have not actually tried it with real lightning payments. That was my next test but i need to fund and open a channel first

![lightning1](https://github.com/user-attachments/assets/af370bbc-dc94-4f8d-8c90-2abeb0d02bde)

![lightning2](https://github.com/user-attachments/assets/da9e22d1-8c39-4e2a-985a-a3681915856b)

**Reply by remcoros** [reply]
ah, I didn't know btcpayserver warns you about this. 👍

**Reply by dennisreimann** [reply]
@mar101xy Also see the documentation on [How to redirect a Lightning Address](https://docs.btcpayserver.org/FAQ/LightningNetwork/#how-to-redirect-a-lightning-address).

**Comment by mar101xy** [comment]
wow! It's possible already? nice! You guys thought of everything. 

That was my original goal. To use a `username@domain.com`

I guess i didn't need to use a sub directory. Others must not know this is possible. I'm going to let everyone know. 

I've been reading all the btcpay docs but haven't got that far yet. The docs are really good.

thanks for the link

**Reply by dennisreimann** [reply]
If it is resolved, can you close the discussion?

**Reply by mar101xy** [reply]
yes. no problem,
I thought it was like forums where it stays open all the time

thank you everyone for the help
